plugins/pulsar-records-storage/src/main/java/com/github/bsideup/liiklus/pulsar/PulsarRecordsStorage.java

```diff
@@ -205,33 +264,71 @@ public boolean equals(Object o) {
-                                                    ByteBuffer.wrap(message.getValue())
-                                            ),
-                                            extractTime(message),
-                                            partition,
-                                            toOffset(message.getMessageId())
-                                    );
-                                })
-                                .delaySubscription(
-                                        initialOffset.flatMap(offset -> {
-                                            return Mono.fromCompletionStage(consumer.seekAsync(fromOffset(offset)));
-                                        })
-                                );
-                    },
-                    consumer -> Mono.fromCompletionStage(consumer.closeAsync())
-            );
+                    })
+                    .doOnNext(__ -> log.debug(
+                            "subscription {}, topic {}, partition {} consumer created",
+                            groupName, topic, partition
+                    ))
+                    .onErrorMap(CompletionException.class, Throwable::getCause)
+                    .doOnError(
+                            e -> !(e instanceof PulsarClientException.ConsumerBusyException),
+                            e -> log.error(
+                                    "subscription {}, topic {}, partition {} failed to create consumer",
+                                    groupName, topic, partition, e
+                            )
+                    )
+                    .doOnError(
+                            e -> e instanceof PulsarClientException.ConsumerBusyException,
+                            e -> log.trace(
+                                    "subscription {}, topic {}, partition {} already connected",
+                                    groupName, topic, partition
+                            )
+                    );
+        }
+
+        private Flux<Record> consumeMessage(Consumer<byte[]> consumer) {
+            return Mono
+                    .fromCompletionStage(consumer::receiveAsync)
+                    .repeat()
+                    .map(message -> {
+                        var key = message.getKey();
+                        return new Record(
+                                new Envelope(
+                                        topic,
+                                        key != null ? ByteBuffer.wrap(key.getBytes()) : null,
+                                        ByteBuffer.wrap(message.getValue())
+                                ),
+                                extractTime(message),
+                                partition,
+                                toOffset(message.getMessageId())
+                        );
+                    })
+                    .delaySubscription(initialOffset.flatMap(offset -> resetSubscriptionOffset(consumer, offset)));
```

Comment:
WDYT if we follow the code that on the current ledger we just don't subtract the offset?
that's maybe not the safest way in the theory since while we are asking about the latest message the ledger could be right after closed, but quick tests are working
 
```
.map(PulsarRecordsStorage::fromOffset)
.flatMap(messageId -> {
        return Mono.fromCompletionStage(() -> ConsumerImplAccessor.getLastMessageIdAsync(consumer))
                         .map(last -> {
                             var msg = (MessageIdImpl) last;

                             if (msg.getLedgerId() == messageId.getLedgerId()) {
                                 return messageId;
                             }

                             return adaptForSeek(messageId);
            });
})
```

@Rowanto @bsideup